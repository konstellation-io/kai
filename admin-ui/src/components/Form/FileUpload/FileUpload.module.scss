@import '../../../styles/mixins';
@import '../../../styles/colors';

.container {
  position: relative;

  &.white,
  &.white .input {
    color: white;
  }

  &.hasClearButton .input {
    padding-right: 30px;
  }
}

.inputContainer {
  display: flex;
  flex-direction: row;
}

.input {
  @include form-input;

  color: $font-color-white;
  font-weight: 600;

  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;

  &.placeholder {
    color: $font-color-regular;
  }
}
