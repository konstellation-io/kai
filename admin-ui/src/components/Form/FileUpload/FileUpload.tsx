import React, { useState, useRef } from 'react';

import InputLabel from '../InputLabel/InputLabel';
import InputError from '../InputError/InputError';
import Button from '../../Button/Button';

import cx from 'classnames';
import styles from './FileUpload.module.scss';

type Props = {
  onChange?: Function;
  placeholder?: string;
  label?: string;
  height?: number;
  error?: string;
};

function FileUpload({
  onChange = function() {},
  placeholder = '',
  label = '',
  height = 40,
  error = ''
}: Props) {
  const [selectedFile, setSelectedFile] = useState();
  const fileButton = useRef(null);

  function onFileUpload() {
    // @ts-ignore
    const file = fileButton.current.files[0];
    setSelectedFile(file);
    onChange(file);
  }

  const inputText = (selectedFile && selectedFile.name) || placeholder;

  return (
    <div className={styles.container}>
      <InputLabel text={label} />
      <div className={styles.inputContainer}>
        <div
          className={cx(styles.input, {
            [styles.placeholder]: !selectedFile,
            [styles.error]: error !== ''
          })}
          style={{ height }}
          data-testid="input"
          title={inputText}
        >
          {inputText}
        </div>
        <Button
          label="BROWSE"
          onClick={() => {
            // @ts-ignore
            fileButton.current.click();
          }}
          border
        />
      </div>
      <input type="file" ref={fileButton} onChange={onFileUpload} hidden />
      <InputError message={error} />
    </div>
  );
}

export default FileUpload;
