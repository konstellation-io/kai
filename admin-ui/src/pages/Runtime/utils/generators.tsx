import React from 'react';
import Button from '../../../components/Button/Button';
import StartIcon from '@material-ui/icons/SkipNext';
import StopIcon from '@material-ui/icons/PauseCircleFilled';
import PublishIcon from '@material-ui/icons/PlayCircleFilledOutlined';
import UnpublishIcon from '@material-ui/icons/Block';
import { VersionStatus } from '../../../graphql/types/globalTypes';

function generateActionButton(label: string, icon: any, action: Function) {
  return (
    <Button
      key={label}
      label={label}
      onClick={action}
      height={30}
      Icon={icon}
    />
  );
}

function getStateToButtons(
  publishAction: Function,
  startAction: Function,
  stopAction: Function,
  unpublishAction: Function
) {
  const buttonStart = generateActionButton('START', StartIcon, startAction);
  const buttonStop = generateActionButton('STOP', StopIcon, stopAction);
  const buttonPublish = generateActionButton(
    'PUBLISH',
    PublishIcon,
    publishAction
  );
  const buttonUnpublish = generateActionButton(
    'UNPUBLISH',
    UnpublishIcon,
    unpublishAction
  );

  return {
    [VersionStatus.STOPPED]: [buttonStart],
    [VersionStatus.PUBLISHED]: [buttonUnpublish],
    [VersionStatus.STARTED]: [buttonPublish, buttonStop],
    [VersionStatus.STOPPED]: [buttonStart]
  };
}

export function getVersionActionButtons(
  publishAction: Function,
  startAction: Function,
  stopAction: Function,
  unpublishAction: Function,
  status?: string
) {
  const stateToButtons: { [key: string]: any } = getStateToButtons(
    publishAction,
    startAction,
    stopAction,
    unpublishAction
  );

  return stateToButtons[status || ''];
}
