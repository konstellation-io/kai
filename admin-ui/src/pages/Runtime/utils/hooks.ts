import { useMutation } from '@apollo/react-hooks';
import {
  ACTIVATE_VERSION,
  DEACTIVATE_VERSION,
  DEPLOY_VERSION,
  STOP_VERSION,
  ActivateVersionResponse,
  DeactivateVersionResponse,
  DeployVersionResponse,
  StopVersionResponse,
  VersionActionVars
} from '../pages/RuntimeStatusPreview/RuntimeStatusPreview.graphql';

export enum versionActions {
  activate = 'activateVersion',
  deploy = 'deployVersion',
  stop = 'stopVersion',
  deactivate = 'deactivateVersion'
}

export default function useVersionAction() {
  const [activateMutation, { loading: loadingM1 }] = useMutation<
    ActivateVersionResponse,
    VersionActionVars
  >(ACTIVATE_VERSION);
  const [deactivateMutation, { loading: loadingM2 }] = useMutation<
    DeactivateVersionResponse,
    VersionActionVars
  >(DEACTIVATE_VERSION);
  const [deployMutation, { loading: loadingM3 }] = useMutation<
    DeployVersionResponse,
    VersionActionVars
  >(DEPLOY_VERSION);
  const [stopMutation, { loading: loadingM4 }] = useMutation<
    StopVersionResponse,
    VersionActionVars
  >(STOP_VERSION);

  const mutationLoading = [loadingM1, loadingM2, loadingM3, loadingM4].some(
    el => el
  );

  function getMutationVars(versionId: string, comment?: string) {
    const variables = {
      variables: {
        input: {
          versionId: versionId
        }
      }
    } as any;

    if (comment) {
      variables.variables.input.comment = comment;
    }

    return variables;
  }

  return {
    [versionActions.activate]: activateMutation,
    [versionActions.deploy]: deployMutation,
    [versionActions.stop]: stopMutation,
    [versionActions.deactivate]: deactivateMutation,
    getMutationVars,
    mutationLoading
  };
}
