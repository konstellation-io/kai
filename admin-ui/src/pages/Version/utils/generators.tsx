import React, { ReactElement, MouseEvent, FunctionComponent } from 'react';
import Button from '../../../components/Button/Button';
import StartIcon from '@material-ui/icons/SkipNext';
import StopIcon from '@material-ui/icons/PauseCircleFilled';
import PublishIcon from '@material-ui/icons/PlayCircleFilledOutlined';
import UnpublishIcon from '@material-ui/icons/Block';
import { VersionStatus } from '../../../graphql/types/globalTypes';
import { SvgIconProps } from '@material-ui/core/SvgIcon';

function generateActionButton(
  label: string,
  icon: FunctionComponent<SvgIconProps>,
  action: (e: MouseEvent<HTMLDivElement>) => void
): ReactElement {
  return (
    <Button
      key={label}
      label={label}
      onClick={action}
      height={30}
      Icon={icon}
    />
  );
}

function getStateToButtons(
  publishAction: (e: MouseEvent<HTMLDivElement>) => void,
  startAction: (e: MouseEvent<HTMLDivElement>) => void,
  stopAction: (e: MouseEvent<HTMLDivElement>) => void,
  unpublishAction: (e: MouseEvent<HTMLDivElement>) => void
): { [key: string]: ReactElement[] } {
  const buttonStart = generateActionButton('START', StartIcon, startAction);
  const buttonStop = generateActionButton('STOP', StopIcon, stopAction);
  const buttonPublish = generateActionButton(
    'PUBLISH',
    PublishIcon,
    publishAction
  );
  const buttonUnpublish = generateActionButton(
    'UNPUBLISH',
    UnpublishIcon,
    unpublishAction
  );

  return {
    [VersionStatus.STOPPED]: [buttonStart],
    [VersionStatus.PUBLISHED]: [buttonUnpublish],
    [VersionStatus.STARTED]: [buttonPublish, buttonStop],
    [VersionStatus.STOPPED]: [buttonStart]
  };
}

export function getVersionActionButtons(
  publishAction: (e: MouseEvent<HTMLDivElement>) => void,
  startAction: (e: MouseEvent<HTMLDivElement>) => void,
  stopAction: (e: MouseEvent<HTMLDivElement>) => void,
  unpublishAction: (e: MouseEvent<HTMLDivElement>) => void,
  status?: string
) {
  const stateToButtons = getStateToButtons(
    publishAction,
    startAction,
    stopAction,
    unpublishAction
  );

  if (status === undefined) {
    return [];
  }

  return stateToButtons[status];
}
