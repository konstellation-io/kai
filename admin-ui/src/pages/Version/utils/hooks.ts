import { loader } from 'graphql.macro';
import { useMutation } from '@apollo/react-hooks';
import {
  GetVersionConfStatus,
  GetVersionConfStatusVariables
} from '../../../graphql/queries/types/GetVersionConfStatus';
import {
  StartVersion,
  StartVersionVariables
} from '../../../graphql/mutations/types/StartVersion';
import {
  StopVersion,
  StopVersionVariables
} from '../../../graphql/mutations/types/StopVersion';
import {
  UnpublishVersion,
  UnpublishVersionVariables
} from '../../../graphql/mutations/types/UnpublishVersion';
import {
  PublishVersion,
  PublishVersionVariables
} from '../../../graphql/mutations/types/PublishVersion';
import { VersionStatus } from '../../../graphql/types/globalTypes';

const PublishVersionMutation = loader(
  '../../../graphql/mutations/publishVersion.graphql'
);
const UnpublishVersionMutation = loader(
  '../../../graphql/mutations/unpublishVersion.graphql'
);
const StartVersionMutation = loader(
  '../../../graphql/mutations/startVersion.graphql'
);
const StopVersionMutation = loader(
  '../../../graphql/mutations/stopVersion.graphql'
);
const GetRuntimeAndVersionsQuery = loader(
  '../../../graphql/queries/getRuntimeAndVersions.graphql'
);

export enum versionActions {
  start = 'startVersion',
  stop = 'stopVersion',
  publish = 'publishVersion',
  unpublish = 'unpublishVersion'
}

export default function useVersionAction(runtimeId: string) {
  const [publishMutation, { loading: loadingM1 }] = useMutation<
    PublishVersion,
    PublishVersionVariables
  >(PublishVersionMutation, {
    update(cache, updateResult) {
      // TODO: This update the previous one activated version.
      // We should remove this when multi activation feature is implemented.
      if (updateResult.data !== undefined && updateResult.data !== null) {
        const updatedVersion = updateResult.data.publishVersion;

        const cacheResult = cache.readQuery<
          GetVersionConfStatus,
          GetVersionConfStatusVariables
        >({
          query: GetRuntimeAndVersionsQuery,
          variables: {
            runtimeId: runtimeId
          }
        });

        if (cacheResult !== null) {
          const { versions, runtime } = cacheResult;
          const newVersions = versions.map(v => {
            if (v.id === updatedVersion.id) return v;
            if (v.status === VersionStatus.PUBLISHED) {
              v.status = VersionStatus.STARTED;
            }
            return v;
          });

          cache.writeQuery({
            query: GetRuntimeAndVersionsQuery,
            data: {
              runtime,
              versions: newVersions
            }
          });
        }
      }
    }
  });

  const [unpublishMutation, { loading: loadingM2 }] = useMutation<
    UnpublishVersion,
    UnpublishVersionVariables
  >(UnpublishVersionMutation);
  const [startMutation, { loading: loadingM3 }] = useMutation<
    StartVersion,
    StartVersionVariables
  >(StartVersionMutation);
  const [stopMutation, { loading: loadingM4 }] = useMutation<
    StopVersion,
    StopVersionVariables
  >(StopVersionMutation);

  const mutationLoading = [loadingM1, loadingM2, loadingM3, loadingM4].some(
    el => el
  );

  type MutationsVariables = {
    variables: PublishVersionVariables;
  };

  function getMutationVars(
    versionId: string,
    comment: string
  ): MutationsVariables {
    return { variables: { input: { versionId, comment } } };
  }

  return {
    [versionActions.publish]: publishMutation,
    [versionActions.start]: startMutation,
    [versionActions.stop]: stopMutation,
    [versionActions.unpublish]: unpublishMutation,
    getMutationVars,
    mutationLoading
  };
}
