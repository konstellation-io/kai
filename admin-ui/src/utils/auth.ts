import Cookies from 'js-cookie';

export const JWT_COOKIE = '__jwt';

export const isUserAuthenticated = () => {
  const jwt = Cookies.get(JWT_COOKIE);

  return jwt;
};
