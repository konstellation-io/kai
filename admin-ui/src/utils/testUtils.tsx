import React from 'react';
import { createMemoryHistory } from 'history';
import { Router, Route } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from '../store';
import { MemoryRouter } from 'react-router';
import { render } from '@testing-library/react';
import { mount } from 'enzyme';
import '@testing-library/jest-dom/extend-expect';

type TestHookprops = {
  callback: Function;
};
const TestHook = ({ callback }: TestHookprops) => {
  callback();
  return null;
};

export const testHook = (callback: Function) => {
  mount(<TestHook callback={callback} />);
};

export function renderWithRouter(
  component: any,
  route: string = '/path/:someParam/index',
  param = { param: 'someParam', value: 'someValue' }
) {
  const history = createMemoryHistory();
  history.push(route.replace(param.param, param.value));

  const element = render(
    <MemoryRouter>
      <Router history={history}>
        <Route path={route}>{component}</Route>
      </Router>
    </MemoryRouter>
  );

  // FIXME: typescript
  // @ts-ignore
  element.customRerender = function(newComponent: any) {
    element.rerender(<MemoryRouter>{newComponent}</MemoryRouter>);
  };

  return {
    element,
    history
  };
}

export function renderWithRedux(component: any, store = configureStore()) {
  return render(<Provider store={store}>{component}</Provider>);
}

export function renderWithReduxAndRouter(
  component: any,
  route = '/path/:someParam/index',
  store = configureStore()
) {
  const renderedElement = renderWithRouter(
    <Provider store={store}>{component}</Provider>,
    route
  );

  // FIXME: typescript definition
  // @ts-ignore
  renderedElement.element.customRerender = function(newComponent: any) {
    renderedElement.element.rerender(
      <Provider store={store}>
        <MemoryRouter>{newComponent}</MemoryRouter>
      </Provider>
    );
  };

  return renderedElement;
}
