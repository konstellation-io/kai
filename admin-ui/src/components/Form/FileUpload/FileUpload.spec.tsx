import React from 'react';
import { renderWithRouter } from '../../../utils/testUtils';
import FileUpload from './FileUpload';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

function renderComponent(params = {}) {
  return renderWithRouter(<FileUpload {...params} />).element;
}

afterEach(cleanup);

it('Render FileUpload without crashing', () => {
  const { container } = renderComponent();

  expect(container).toMatchSnapshot();
});

it('show right texts', () => {
  const { getByText } = renderComponent({
    placeholder: 'some placeholder',
    label: 'some label'
  });

  expect(getByText('some placeholder')).toBeInTheDocument();
  expect(getByText('SOME LABEL')).toBeInTheDocument();
});
