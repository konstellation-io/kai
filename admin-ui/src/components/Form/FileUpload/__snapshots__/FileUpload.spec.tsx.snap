// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render FileUpload without crashing 1`] = `
<div>
  <div
    class="container"
  >
    <label
      class="label"
      data-testid="label"
    />
    <div
      class="inputContainer"
    >
      <div
        class="input placeholder"
        data-testid="input"
        style="height: 40px;"
        title=""
      />
      <div
        class="btn default middle border label"
        style="height: 40px; line-height: 40px;"
      >
        <span>
          BROWSE
        </span>
      </div>
    </div>
    <input
      hidden=""
      type="file"
    />
    <div
      class="errorMessage"
      data-testid="error-message"
    />
  </div>
</div>
`;
